system-tests-new/tiny-project/test_tiny_project.py

```diff
@@ -1,56 +1,217 @@
+            path_filters=[Exclude('log.txt', '.csv', '/metrics/', build_config.source_root)])
+
+        hashes = {}
+        for f in output_files:
+            hashes[f] = file_checksum(f).file_hash
+
+        return hashes
+
+    def get_csvs(self, build_config):
+        """
+        Get the contents of each csv file in the project workspace.
+
+        The row order is not deterministic so we put rows in a set for easy comparison.
+        Each row is a dict, which can't go in a set, so we sort and tuple row items for easy comparison.
+
+        """
+        output_files = self.find_output_files(
+            root=build_config.project_workspace,
+            path_filters=[
+                Exclude(build_config.project_workspace),
+                Include('.csv'),
+            ])
+
+        csvs = {}
+        for f in output_files:
+            csvs[f] = set([tuple(sorted(row.items())) for row in csv.DictReader(open(f))])
+
+        return csvs
+
+    def find_output_files(self, root, path_filters):
+        # might as well use this build step to find the output files, as it already does what we need
+        file_finder = FindSourceFiles(
+            source_root=root,
+            path_filters=path_filters,
+            output_collection='output files',
+        )
+
+        temp_store = {}
+        file_finder.run(artefact_store=temp_store, config=None)
+
+        return temp_store['output files']
+
+    def test_no_change_rebuild(self, configs):
+        # ensure a rebuild with no changes does not recompile any fortran
+        grab_config, build_config = configs
+        grab_config.run()
+
+        # clean build
+        clean_timestamps, clean_hashes, clean_csvs = self.build(build_config)
+
+        # rebuild
+        rebuild_timestamps, rebuild_hashes, rebuild_csvs = self.build(build_config)
+
+        # make sure no Fortran compilation output has been recreated
+        fortran_timestamps = set(filter(lambda i: i[0].suffix in ['.o', '.mod'], rebuild_timestamps.items()))
+        assert fortran_timestamps <= set(clean_timestamps.items())
+        assert rebuild_hashes == clean_hashes
+        assert rebuild_csvs == clean_csvs
+
+    def test_incremental_build_no_mod(self, configs):
+        # test incremental fortran build, no module change
+        grab_config, build_config = configs
+        grab_config.run()
+
+        # clean build
+        clean_timestamps, clean_hashes, clean_csvs = self.build(build_config)
+
+        # modify the fortran module source without changing the module interface
+        mod_source = build_config.source_root / 'src/my_mod.F90'
+        lines = open(mod_source, 'rt').readlines()
+        with open(mod_source, 'wt') as out:
+            for line in lines:
+                out.write(line)
+                # duplicate the print line
+                if 'PRINT' in line:
+                    out.write(line)
+
+        # rebuild
+        rebuild_timestamps, rebuild_hashes, _ = self.build(build_config)
```

Comment:
Why do we not care about the CSV file here? Particularly when we did for the clean build.